runtime/binding-http/src/main/java/io/aklivity/zilla/runtime/binding/http/internal/config/HttpBindingConfig.java

```diff
@@ -193,57 +194,65 @@ private Function<Function<String, String>, String> asAccessor(
+                Map<String8FW, Validator> headers = new HashMap<>();
                 if (request.headers != null)
                 {
-                    requestType.headers = new HashMap<>();
                     for (HttpParamConfig header : request.headers)
                     {
-                        String8FW name = new String8FW(header.name);
-                        requestType.headers.put(name, createValidator.apply(header.validator, this.resolveId));
+                        headers.put(new String8FW(header.name), createValidator.apply(header.validator, this.resolveId));
                     }
                 }
+                Map<String, Validator> pathParams = new HashMap<>();
                 if (request.pathParams != null)
                 {
-                    requestType.pathParams = new HashMap<>();
                     for (HttpParamConfig pathParam : request.pathParams)
                     {
-                        requestType.pathParams.put(pathParam.name, createValidator.apply(pathParam.validator, this.resolveId));
+                        pathParams.put(pathParam.name, createValidator.apply(pathParam.validator, this.resolveId));
                     }
                 }
+                Map<String, Validator> queryParams = new TreeMap<>(this::urlDecodedComparator);
                 if (request.queryParams != null)
                 {
-                    requestType.queryParams = new HashMap<>();
                     for (HttpParamConfig queryParam : request.queryParams)
                     {
-                        requestType.queryParams.put(queryParam.name, createValidator.apply(queryParam.validator, this.resolveId));
+                        queryParams.put(queryParam.name, createValidator.apply(queryParam.validator, this.resolveId));
                     }
                 }
-                if (request.content != null)
-                {
-                    requestType.content = createValidator.apply(request.content, this.resolveId);
-                }
+                Validator content = createValidator.apply(request.content, this.resolveId);
+                HttpRequestType requestType = HttpRequestType.builder()
+                    .path(request.path)
+                    .method(request.method)
+                    .contentType(request.contentType)
+                    .headers(headers)
+                    .pathParams(pathParams)
+                    .queryParams(queryParams)
+                    .content(content)
+                    .build();
                 result.add(requestType);
             }
         }
         return result;
     }
 
-    public HttpRequestType resolveRequest(
+    // visible for testing
+    int urlDecodedComparator(
+        String input1,
+        String input2)
+    {
+        String decoded1 = URLDecoder.decode(input1, UTF_8);
+        String decoded2 = URLDecoder.decode(input2, UTF_8);
+        return decoded1.compareTo(decoded2);
```

Comment:
Please extract this into a separate class, such as `PercentEncodableStringComparator` that implements `Comparator<String, String>`.

The purpose of the implementation is to avoid allocations, just readonly operations to compare. This implementation does not do that yet.

Instead, we want to walk down each string in parallel, char by char, advancing either 1 char (no URL encoding) or 3 chars (% plus 2 chars) for each, until we get an effective mismatch of char vs char, we discover effective decoded length mismatch, or both strings are equivalent.

Then no allocation will occur during compare.

Reply:
OK, pls chk.